### PR TITLE
Better error handling of terminated channels

### LINKED ISSUE #3313: Deleted Channels Throw An Error When Refreshing Subscription Feed
Using the latest version. If the user imports a subscription list with a deleted chaneel such as 
image
results in a null pointer exception as a snack-bar error.

## Exception
* __User Action:__ requested feed
* __Request:__ Loading feed
* __Content Language:__ US
* __Service:__ none
* __Version:__ 0.19.0
* __OS:__ Linux Android 10 - 29

<details><summary><b>Crash log</b></summary><p>

```
org.schabi.newpipe.local.feed.service.FeedLoadService$RequestException: 0:https://www.youtube.com/channel/UCcNE9TRcOCOoOB0Dy5ediPg
	at org.schabi.newpipe.local.feed.service.FeedLoadService$startLoading$7.apply(FeedLoadService.kt:220)
	at org.schabi.newpipe.local.feed.service.FeedLoadService$startLoading$7.apply(FeedLoadService.kt:63)
	at io.reactivex.internal.operators.parallel.ParallelMap$ParallelMapSubscriber.onNext(ParallelMap.java:113)
	at io.reactivex.internal.operators.parallel.ParallelFilter$ParallelFilterSubscriber.tryOnNext(ParallelFilter.java:130)
	at io.reactivex.internal.operators.parallel.ParallelRunOn$RunOnConditionalSubscriber.run(ParallelRunOn.java:397)
	at io.reactivex.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:66)
	at io.reactivex.internal.schedulers.ScheduledRunnable.call(ScheduledRunnable.java:57)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:301)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
	at java.lang.Thread.run(Thread.java:923)
Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'java.lang.String org.jsoup.nodes.Element.text()' on a null object reference
	at org.schabi.newpipe.extractor.services.youtube.extractors.YoutubeFeedExtractor.getId(YoutubeFeedExtractor.java:53)
	at org.schabi.newpipe.extractor.feed.FeedInfo.getInfo(FeedInfo.java:39)
	at org.schabi.newpipe.util.ExtractorHelper.lambda$getFeedInfoFallbackToChannelInfo$6(ExtractorHelper.java:149)
	at org.schabi.newpipe.util.-$$Lambda$ExtractorHelper$ugdV6CfeDCyyO_ocIIpHyBlriuE.call(Unknown Source:4)
	at io.reactivex.internal.operators.maybe.MaybeFromCallable.subscribeActual(MaybeFromCallable.java:46)
	at io.reactivex.Maybe.subscribe(Maybe.java:4154)
	at io.reactivex.internal.operators.maybe.MaybeSwitchIfEmptySingle.subscribeActual(MaybeSwitchIfEmptySingle.java:45)
	at io.reactivex.Single.subscribe(Single.java:3438)
	at io.reactivex.Single.blockingGet(Single.java:2699)
	at org.schabi.newpipe.local.feed.service.FeedLoadService$startLoading$7.apply(FeedLoadService.kt:210)
	... 11 more

```
</p></details>
<hr>

### LINKED ISSUE #5452: Feed fetching errors has bad UI/UX
### Checklist

- [x] I am using the latest version
- [x] I checked, but didn't find any duplicates (open OR closed) of this issue in the repo
- [x] I have read the contribution guidelines given at https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md.
- [x] This issue contains only one bug. I will open one issue for every bug report I want to file.

### Steps to reproduce the bug
1. Open Feed
2. Pull down to update
3. Some feed(s) fail to update

### Actual behaviour
Network Error - Retry
No videos are shown, even if 99% of feeds successfully fetched.
Some PeerTube feeds repeatedly fail, probably because of enterprise firewall and alike.

### Expected behavior
Successfully retrieved feeds/videos are shown
The error would show on top of the list like: "xx out of yy channels failed to update. Pull down again to retry"

### PR DESCRIPTION
**Pull request by TobiGr**
#### What is it?
- [x] Bugfix (user facing)
- [ ] Feature (user facing)
- [ ] Codebase improvement (dev facing)
- [ ] Meta improvement to the project (dev facing)

#### Description of the changes in your PR
- give a reason, why a feed could not be loaded when possible
- allow to unsubscribe from the dedicated channel
- show that an account has been terminated and (if provided) the reason for the termination in ChannelFragment

#### ToDo
- [x] display the information in the channel fragment
- [x] loading state is not changed correctly for some reason.

#### Fixes the following issue(s)

- Fixes #3313 
- Closes #5452

#### Relies on the following changes

- TeamNewPipe/NewPipeExtractor#580

#### Screenshots
<details><summary>Screenshots</summary>

</details>

#### APK testing 

On the website the APK can be found by going to the "Checks" tab below the title and then on "artifacts" on the right.

I manipulated the subscription export to contain only URls to termianted or not available channels. Extract the json and import it for testing:
[newpipe_termianted.json.zip](https://github.com/TeamNewPipe/NewPipeExtractor/files/6239718/newpipe_termianted.json.zip)
#### Due diligence
- [x] I read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by opusforlife2** [comment]
Sorely needed!

But... how do you test? xD

File: app/src/main/res/values/strings.xml [hunk 1]
```diff
@@ -688,11 +688,16 @@
...
     <string name="feed_use_dedicated_fetch_method_enable_button">Enable fast mode</string>
     <string name="feed_use_dedicated_fetch_method_disable_button">Disable fast mode</string>
     <string name="feed_use_dedicated_fetch_method_help_text">Do you think feed loading is too slow? If so, try enabling fast loading (you can change it in settings or by pressing the button below).\n\nNewPipe offers two feed loading strategies:\n• Fetching the whole subscription channel, which is slow but complete.\n• Using a dedicated service endpoint, which is fast but usually not complete.\n\nThe difference between the two is that the fast one usually lacks some information, like the item\'s duration or type (can\'t distinguish between live videos and normal ones) and it may return less items.\n\nYouTube is an example of a service that offers this fast method with its RSS feed.\n\nSo the choice boils down to what you prefer: speed or precise information.</string>
+
```
Comment (mhmdanas) [inline app/src/main/res/values/strings.xml:700]:
```suggestion
```

File: app/src/main/java/org/schabi/newpipe/local/feed/service/FeedLoadService.kt [hunk 2]
```diff
@@ -345,6 +347,17 @@ class FeedLoadService : Service() {
                     error is IOException -> throw error
                     cause is IOException -> throw cause
                     error.isNetworkRelated -> throw IOException(error)
+
+                    cause is ContentNotAvailableException -> {
+                        // maybeWrapper is definitely a RequestException,
+                        // because this is an exception thrown in the extractor
+                        if (maybeWrapper is RequestException) {
+                            Log.d(TAG, if (error.message == null) "terminated" else error.message)
```
Comment (mhmdanas) [inline app/src/main/java/org/schabi/newpipe/local/feed/service/FeedLoadService.kt:355]:
```suggestion
                            Log.d(TAG, error.message ?: "terminated")
```

**Review by mhmdanas — CHANGES_REQUESTED** [review]
Two minor things. I guess the files should probably be formatted too.

File: app/src/main/res/values/strings.xml (hunk 1 above)
Comment (TobiGr) [reply app/src/main/res/values/strings.xml:701]: That blank line was intended to separate the feed strings from the others

File: app/src/main/java/org/schabi/newpipe/local/feed/service/FeedLoadService.kt (hunk 2 above)
Comment (TobiGr) [reply app/src/main/java/org/schabi/newpipe/local/feed/service/FeedLoadService.kt]: Removed. I did not want to add that debug message

**Comment by TobiGr** [comment]
> But... how do you test? xD

@opusforlife2  I added  a manipulated subscription export to the PR description

@Stypox Can you help me with the feed loading UI? I am unable to solve that at the moment. Just install the APK, import the subscriptions, refresh the feed and hit unsubscribe or cancel on error to see the problem.

File: app/src/main/res/values/strings.xml [hunk 3]
```diff
@@ -708,6 +713,8 @@
     <string name="soundcloud_go_plus_content">This is a SoundCloud Go+ track, at least in your country, so it cannot be streamed or downloaded by NewPipe.</string>
     <string name="private_content">This content is private, so it cannot be streamed or downloaded by NewPipe.</string>
     <string name="youtube_music_premium_content">This video is available only to YouTube Music Premium members, so it cannot be streamed or downloaded by NewPipe.</string>
+    <string name="account_terminated">Account terminated</string>
+    <string name="service_provides_reason">%s provides this reason:</string>
```
Comment (Stypox) [inline app/src/main/res/values/strings.xml:717]: Maybe some language would put the message in another place
```suggestion
    <string name="service_provides_reason">%1$s provides this reason: %2$s</string>
```

**Review by Stypox — CHANGES_REQUESTED** [review]
I am not sure I understand what I should test. When importing the subscriptions you provided nothing happens (i.e. no subscriptions are added). I tried both importing the json from the subscription fragment and importing the zip from content settings

File: app/src/main/res/values/strings.xml (hunk 3 above)
Comment (TobiGr) [reply app/src/main/res/values/strings.xml:717]: That messages is in another TextView with extra margin and italic fonts.

Comment (Stypox) [reply app/src/main/res/values/strings.xml:717]: Oh right, sorry

**Comment by TobiGr** [comment]
@Stypox Did you extract the json from the ZIP file before importing it via the subscription import dialog?

Anyway, here's a screen cast of the problem. As you can see, the feed keeps loading forever, although loading the items is completed.


https://user-images.githubusercontent.com/17365767/113301460-54165e80-92ff-11eb-9c09-dd42781e5f21.mp4

**Comment by Stypox** [comment]
@TobiGr I tried to understand how the feed loading works but I have failed so far, will try again tomorrow.

**Comment by Stypox** [comment]
@TobiGr @opusforlife2 what do you think of this implementation? It now shows correctly the loaded feed and shows a dialog for **each** invalid subscription. If something unrelated to channel being terminated goes wrong (e.g. a network error), the bar at the top correctly shows the number of channels that could not be loaded, instead of showing "Network error". Here is a *settings* export (--> import it from settings): [NewPipeData-20210402_120407.zip](https://github.com/TeamNewPipe/NewPipe/files/6250379/NewPipeData-20210402_120407.zip)

**Comment by TobiGr** [comment]
Thank you. Looks good!

I fixed two small bugs (see commit message)

**Comment by TobiGr** [comment]
[NewPipeData-20210402_120407.zip](https://github.com/TeamNewPipe/NewPipe/files/6251211/NewPipeData-20210402_120407.zip)
The Pokemon channel is terminated now.

**Comment by opusforlife2** [comment]
Sorry, this slipped my mind yesterday. From the Subscriptions tab, I can see the 2 kinds of errors you pointed out: content unavailable, and channel terminated.

Refreshing a feed with the terminated Pokemon channel shows "channel terminated, wanna unsub?" so it's fine.

Doing the same with the content unavailable channel shows "got error: channel does not exist." No idea if the message is correct here. "Content unavailable" and "channel does not exist" seem like two different things.

Now the bugs:

- If you put both the terminated and content unavailable channels in a channel group and refresh the feed, or try refreshing the What's New feed, you get one of the two dialogs, randomly selected. If you unsubscribe from one, you are shown the other as well. But if you tap cancel or tap outside the dialog box to dismiss it, you are not shown the second exception. It should show both each time.
- Once you get either exception, cancel and go back to the Subscriptions tab. You will encounter the last shown exception as you back out of the channel group. Now keep opening one of the two valid channels and tapping Back. You keep getting the last stored exception every time you return to the Subscriptions tab upon tapping Back.
- But if you do the same thing after unsubscribing from at least one or both channels, you don't get an exception with the same steps.

**Comment by TobiGr** [comment]
Rebased

**Review by Stypox — APPROVED** [review]
LGTM, thank you :-)

**Comment by opusforlife2** [comment]
Did you fix the bugs I found?

**Pull request merged**